Initial support of Dreo heaters

andy-cooper opened the pull request:
I modeled this as a Home Assistant climate device so that we can get some nice graphical thermostats. I did a little bit of refactoring of existing code to try to fit things into the same framework and make it a little more general - that work is incomplete. 

This is largely functional against the one type of heater that I have - looking at Dreo's website, the other type of heaters look very similar, so I suspect that they should generally work if added to the supported models list.

In the spirit of giving, I thought I'd try to contribute this back to the community. Many thanks, @JeffSteinbok, for the work you did to get this off the ground. I doubt I would have tried to do this were it not for your earlier work.

andy-cooper commented:
Don't merge this yet - it looks like there's an issue with it taking up 100% CPU of the homeassistant container. I need to debug it further ...

andy-cooper commented:
OK, looks like I've fixed it. Go ahead and feel free to merge or not 😃

andy-cooper commented:
Also, rather than double-posting the images - take a look at what I put into https://github.com/JeffSteinbok/hass-dreo/issues/61 to see the UX.

JeffSteinbok commented on README.md:
```diff
@@ -42,8 +42,58 @@ Currently supported models are listed below.
+
+<table>
+    <tr>
+        <td><img src="https://raw.githubusercontent.com/andy-cooper/hass-dreo/d1e67331ba1be79a8635a6c3aaa54ac279e7a243/auto-mode-thermostat.png" width="300" alt="eco/auto mode thermostat"></td>
```
Please ensure the URLs are correct after the merge.

andy-cooper replied:
Yup. I *thought* I had done those as relative URLs. I'll make sure they're correct after merge.

andy-cooper replied:
Ah - I know why they're not relative - relative links apparently don't work. Once merged, we'll need to change the links.

JeffSteinbok commented on custom_components/dreo/__init__.py:
```diff
@@ -45,34 +46,51 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
     platforms = []
 
-    _LOGGER.debug(device_dict)
+    _LOGGER.debug("Device dict: %s", device_dict)
```
device_dict isn't modified is it?  Why log again?

andy-cooper replied:
Detritus from an earlier debugging attempt. Removed.

JeffSteinbok commented on custom_components/dreo/climate.py:
```diff
@@ -0,0 +1,350 @@
+    ECOLEVEL_RANGE
+)
+
+from homeassistant.components.climate import (
```
Please consider moving these into .haimports.  I had them all there to avoid some collisions AFAIR

andy-cooper replied:
Will do.

andy-cooper replied:
Done.

JeffSteinbok commented on custom_components/dreo/climate.py:
```diff
@@ -0,0 +1,350 @@
+    _attr_temperature_unit = UnitOfTemperature.FAHRENHEIT # TODO
+    _attr_target_temperature = None
+    _attr_current_temperature = None
+    _attr_fan_mode = None
```
Is it worth refactoring Fan/Heater so they share more code?

andy-cooper replied:
At some point yes, but initially I wanted to keep them separate and not touch the Fan code too much because I didn't want to break anything. Not having a fan myself, I thought it best if I keep it separate for the moment.

JeffSteinbok commented on custom_components/dreo/switch.py:
```diff
@@ -103,8 +145,9 @@ def __init__(self,
         """Return True if device is on."""
-        _LOGGER.debug("DreoSwitchHA:is_on")
-        return getattr(self.device, self.entity_description.attr_name)
+        attr = getattr(self.device, self.entity_description.attr_name)
```
Why remove the log line?

andy-cooper replied:
Accident. Fixed.

JeffSteinbok commented on hacs.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "Dreo Smart Device Integration",
+    "name": "Dreo Smart Device Integration with Heaters",
```
Remove.  Heaters are Smart Devices.

andy-cooper replied:
That shouldn't have been in the diffs - I had done it just to differentiate my fork in a list. It's not going to go into the merge.

andy-cooper replied:
Removed.

JeffSteinbok commented on custom_components/dreo/pydreo/pydreoheater.py:
```diff
@@ -0,0 +1,414 @@
+class PyDreoHeater(PyDreoBaseDevice):
+    """Base class for Dreo heater API Calls."""
+
+    def __init__(self, device_definition: DreoDeviceDetails, details: Dict[str, list], dreo: "PyDreo"):
```
How much of this is duplicative of what's in Fan?

andy-cooper replied:
The structure is the same, but the methods and properties are largely different. There are some common things that should at some point be refactored out into the base device. But as I noted, I wanted to keep things as similar to the old code as possible to avoid breaking the fans because I don't have one and can't easily test it.

andy-cooper commented:
@JeffSteinbok - I think I've addressed all the comments. LMK what you think at this point.

JeffSteinbok commented:
Couple quick things, then good to go.

andy-cooper commented:
> Couple quick things, then good to go.

I don't see any more change requests than the ones I addressed (and the 2 about refactoring that I made a note on). All other things I fixed and resolved the conversation. Did you make extra comments/requests after that that I missed?

andy-cooper commented:
@JeffSteinbok - can you take another look here; I think I addressed everything.

andy-cooper commented:
Thanks!! I've got some more changes for oscillation support for heaters that I'll make a PR for in the next few days. And then I'll take a look at refactoring some stuff.
